File: exercism/javascript/binary-search/binary-search.js

const ascending = (xs) =>
  xs.reduce(([b, x], y) => [b && x <= y, y], [true, -Infinity])[0]

class BinarySearch {
  constructor (xs) {
    if (ascending(xs)) this.array = xs
  }
  indexOf (x) {
    let [i, j] = [-1, this.array.length]
    while (i < j) {
      const mid = Math.floor((i + j) / 2)
      const probe = this.array[mid]
      if (x === probe) return mid
      if (x < probe) {
        j = mid - 1
      } else {
        i = mid + 1
      }
    }
    return -1
  }
}

module.exports = BinarySearch

File: exercism/python/binary-search/binary_search.py

def binary_search(xs, n):
    a, b = 0, len(xs) - 1
    while (a <= b):
        probe = (a + b) / 2
        candidate = xs[probe]
        if candidate == n:
            return probe
        if candidate < n:
            a = probe + 1
        else:
            b = probe - 1
    raise ValueError
